alacritty_terminal/src/text_run.rs

```diff
@@ -0,0 +1,289 @@
+            span: (start.column, end_column),
+            content: TextRunContent::CharRun(buffer.0, buffer.1),
+            fg: start.fg,
+            bg: start.bg,
+            bg_alpha: start.bg_alpha,
+            flags: start.flags,
+        }
+    }
+
+    fn from_cursor_rc(start: RunStart, cursor: CursorKey) -> Self {
+        TextRun {
+            line: start.line,
+            span: (start.column, start.column),
+            content: TextRunContent::Cursor(cursor),
+            fg: start.fg,
+            bg: start.bg,
+            bg_alpha: start.bg_alpha,
+            flags: start.flags,
+        }
+    }
+
+    /// Holdover method while converting from rendering Cells to TextRuns
+    fn cell_at(&self, col: Column) -> RenderableCell {
+        RenderableCell {
+            line: self.line,
+            column: col,
+            inner: RenderableCellContent::Chars([' '; crate::term::cell::MAX_ZEROWIDTH_CHARS + 1]),
+            fg: self.fg,
+            bg: self.bg,
+            bg_alpha: self.bg_alpha,
+            flags: self.flags,
+        }
+    }
+
+    /// First cell in the TextRun
+    pub fn start_cell(&self) -> RenderableCell {
+        self.cell_at(self.span.0)
+    }
+
+    /// First point covered by this TextRun
+    pub fn start_point(&self) -> Point {
+        Point { line: self.line, col: self.span.0 }
+    }
+
+    /// End point covered by this TextRun
+    pub fn end_point(&self) -> Point {
+        Point { line: self.line, col: self.span.1 }
+    }
+
+    /// Iterates over each RenderableCell in column range [run.0, run.1]
+    pub fn cells<'a>(&'a self) -> impl Iterator<Item = RenderableCell> + 'a {
+        let step = if self.flags.contains(Flags::WIDE_CHAR) { 2 } else { 1 };
+        let (Column(start), Column(end)) = self.span;
+        // TODO: impl Step for Column (once Step is stable) to avoid unwrapping then rewrapping.
+        (start..=end).step_by(step).map(move |col| self.cell_at(Column(col)))
+    }
+}
+
+type IsWide = bool;
+type LatestCol = (Column, IsWide);
+
+/// Wraps an Iterator<Item=RenderableCell> and produces TextRuns to represent batches of cells
+pub struct TextRunIter<I> {
+    iter: I,
+    run_start: Option<RunStart>,
+    latest_col: Option<LatestCol>,
+    cursor: Option<CursorKey>,
+    buffer_text: String,
+    buffer_zero_width: Vec<[char; MAX_ZEROWIDTH_CHARS]>,
+}
+
+// This is an explicit function (as opposed to a closure) to make it easy to use as a function
+// pointer.
+fn is_not_wide_char_spacer(rc: &RenderableCell) -> bool {
+    !rc.flags.contains(Flags::WIDE_CHAR_SPACER)
+}
+impl<BaseIter> TextRunIter<std::iter::Filter<BaseIter, fn(&RenderableCell) -> bool>>
+where
+    BaseIter: Iterator<Item = RenderableCell>,
+{
+    pub fn new(iter: BaseIter) -> Self {
+        TextRunIter {
+            // Logic for WIDE_CHAR is handled internally by TextRun
+            // So we no longer need WIDE_CHAR_SPACER at this point.
+            iter: iter.filter(is_not_wide_char_spacer),
```

Comment:
I think we can. The rationale for putting it here is because TextRunIter is the one that cares about filtering out the WIDE_CHAR_SPACER characters. While there currently aren't any, in theory any other consumer of the renderable_cells() method would want the full stream of RenderableCells unfiltered.

That being said RenderableCellsIter and TextRunIter are very tightly coupled and I think they could more aggressively be combined. For example, we could build TextRuns directly from `Cell` and avoid going through RenderableCellsIter entirely. I haven't taken a look at grid code at all though so I have don't know how involved that would be.

Reply:
As far as I can tell, we're currently using the TextRun on all platforms anyways and it's a fixed part of our pipeline. So I don't see why we wouldn't integrate it properly.

We only need the renderable cells to render text. If we always render text through the text run iter, we should be fine with reworking the pipeline to fit us as well as possible.